soh/soh/Enhancements/crowd-control/CrowdControl.cpp

```diff
@@ -19,268 +21,301 @@ extern GlobalContext* gGlobalCtx;
+    ccThreadProcess = std::thread(&CrowdControl::ProcessActiveEffects, this);
 }
 
-void CrowdControl::RunCrowdControl(CCPacket* packet) {
-    bool paused = false;
-    EffectResult lastResult = EffectResult::NotReady;
-    bool isTimed = packet->timeRemaining > 0;
-
-    while (connected) {
-        EffectResult effectResult = ExecuteEffect(packet->effectType.c_str(), packet->effectValue, 0);
-        if (effectResult == EffectResult::Success) {
-            // If we have a success after being previously paused, we fire the Resumed event
-            if (paused && packet->timeRemaining > 0) {
-                paused = false;
-                nlohmann::json dataSend;
-                dataSend["id"] = packet->packetId;
-                dataSend["type"] = 0;
-                dataSend["timeRemaining"] = packet->timeRemaining;
-                dataSend["status"] = EffectResult::Resumed;
-
-                std::string jsonResponse = dataSend.dump();
-                SDLNet_TCP_Send(tcpsock, jsonResponse.c_str(), jsonResponse.size() + 1);
-            }
+void CrowdControl::Disable() {
+    if (!isEnabled) {
+        return;
+    }
 
-            // If time remaining has reached 0 or was 0, we have finished let's remove the command and end the thread
-            if (packet->timeRemaining <= 0) {
-                receivedCommandsMutex.lock();
-                receivedCommands.erase(std::remove(receivedCommands.begin(), receivedCommands.end(), packet), receivedCommands.end());
-                receivedCommandsMutex.unlock();
-                RemoveEffect(packet->effectType.c_str());
-
-                // If not timed, let's fire the one and only success
-                if (!isTimed) {
-                    nlohmann::json dataSend;
-                    dataSend["id"] = packet->packetId;
-                    dataSend["type"] = 0;
-                    dataSend["timeRemaining"] = packet->timeRemaining;
-                    dataSend["status"] = EffectResult::Success;
-
-                    std::string jsonResponse = dataSend.dump();
-                    SDLNet_TCP_Send(tcpsock, jsonResponse.c_str(), jsonResponse.size() + 1);
-
-                    delete packet;
-                }
+    isEnabled = false;
+    ccThreadReceive.join();
+    ccThreadProcess.join();
+}
+
+void CrowdControl::ListenToServer() {
+    while (isEnabled) {
+        while (!connected) {
+            SPDLOG_TRACE("[CrowdControl] Attempting to make connection to server...");
+            tcpsock = SDLNet_TCP_Open(&ip);
 
-                return;
+            if (tcpsock) {
+                connected = true;
+                SPDLOG_TRACE("[CrowdControl] Connection to server established!");
+                break;
             }
+        }
+
+        // Listen to socket messages
+        while (connected && tcpsock && isEnabled) {
+            auto socketSet = SDLNet_AllocSocketSet(1);
+            SDLNet_TCP_AddSocket(socketSet, tcpsock);
```

Comment:
This should be moved out of the loop. There is no need to allocate a new set every time.